nncf/common/tensor_statistics/collectors.py

```diff
@@ -0,0 +1,280 @@
+
+    @abstractmethod
+    def _get_statistics(self):
+        pass
+
+    def enable(self):
+        self._enabled = True
+
+    def disable(self):
+        self._enabled = False
+
+    def reset(self):
+        self._collected_samples = 0
+        self._reset()
+
+    @abstractmethod
+    def _reset(self):
+        pass
+
+    def collected_samples(self) -> int:
+        return self._collected_samples
+
+
+class StatisticsNotCollectedError(Exception):
+    """Raised when the statistics are not collected but requested"""
+    pass
+
+
+class OnlineTensorStatisticCollector(TensorStatisticCollectorBase, ABC):
+    pass
+
+
+class OfflineTensorStatisticCollector(TensorStatisticCollectorBase, ABC):
+    def __init__(self, reduction_shape: ReductionShape = None, num_samples: int = None, window_size: int = None):
+        super().__init__(reduction_shape, num_samples)
+        self._samples = deque(maxlen=window_size)
+
+    def _reset(self):
+        self._samples.clear()
+
+
+class CollectorParams:
+    def __init__(self, is_weights: bool, mode: str, per_channel: bool, init_type: str):
+        self._is_weights = is_weights
+        self._mode = mode
+        self._per_channel = per_channel
+        self._init_type = init_type
+
+    @property
+    def _use_abs_max(self) -> bool:
+        return self._mode == 'symmetric'
+
+    @property
+    def _use_means_of_mins(self) -> bool:
+        return not self._is_weights and not self._per_channel and self._mode == 'asymmetric'
+
+    @property
+    def _use_means_of_maxs(self) -> bool:
+        return not self._is_weights and not self._per_channel
+
+    def _get_params_for_min_max(self) -> bool:
+        return self._use_abs_max
+
+    def _get_params_for_mixed_min_max(self) -> Tuple[bool, bool, bool]:
+        return self._use_abs_max, self._use_means_of_mins, self._use_means_of_maxs
+
+    def _get_params_for_mean_min_max(self) -> bool:
+        return self._use_abs_max
+
+    def get_low_level_params_for_collector(self):
+        """Generate low-level parameters for collectors"""
+        if self._init_type == "min_max":
+            return self._get_params_for_min_max()
+        if self._init_type == "mixed_min_max":
+            return self._get_params_for_mixed_min_max()
+        if self._init_type == "mean_min_max":
+            return self._get_params_for_mean_min_max()
+        raise RuntimeError("Parameters not required or unknown range init type: {}".format(self._init_type))
+
+
+class MinMaxStatisticCollector(OnlineTensorStatisticCollector, ABC):
+    """
+    Collector estimates min of minimum values and max of maximum values.
+    """
+
+    def __init__(self, use_abs_max: bool, reduction_shape: ReductionShape,
+                 num_samples: int = None, nncf_backend = None):
+        super().__init__(reduction_shape, num_samples)
+        self._use_abs_max = use_abs_max
+
+        self._min_values = None
+        self._max_values = None
+
+        if nncf_backend == BackendType.TORCH:
```

Comment:
This probably should not be determined via an init parameter. Why not use inheritance and polymorphism, implementing corresponding aggregation functions in framework-specific inheriting classes?

Reply:
Thanks. I redefined it using inheritance and polymorphism.